
$\qquad$ $\qquad$ $\qquad$  **TDA231/DIT370 Discrete Optimization: Home Assignment 3 -- LP Duality and the Primal-Dual Algorithm** <br />
$\qquad$ $\qquad$ $\qquad$                   **During grading time, direct queries regarding Q-1,2 to David & Q-3 to Marc** <br />
$\qquad$ $\qquad$ $\qquad$                     **Due Date: 03/03/2025** <br />
$\qquad$ $\qquad$ $\qquad$                   **Submitted by: Name, Personal No., Email** <br />



---


General guidelines:
*   All solutions to theoretical and pratical problems must be submitted in this ipynb notebook and equations, wherever required, should be formatted using LaTeX math-mode.
*   All discussion regarding practical problems, along with solutions and plots should be specified in this notebook. All plots/results should be visible such that the notebook does not have to be run. But the code in the notebook should reproduce the plots/results if we choose to do so.
*   Your name, personal number and email address should be specified above.
*   All tables and other additional information should be included in this notebook.
*   Before submitting, make sure that your code can run on another computer. That all plots can show on another computer including all your writing. It is good to check if your code can run here: https://colab.research.google.com.


# Problem 1.

Consider the following LP problem:

\begin{alignat*}{2}
\max \ &4x_1-2x_2+5x_3+6x_4+7x_5\\
\\
\textrm{s.t} \quad
&2x_1 + 2x_2 - 4x_3 + 4x_4 + 8x_5 &&\leq 6\\
&2x_1 + \ \ {}x_2 - 2x_3 - \ \ x_4 - 3x_5 &&\geq -1\\
&5x_1 - 2x_2 + 4x_3 + 4x_4 + 2x_5 &&= 5\\
&2x_1 - 2x_2 + 5x_3 + 3x_4 + \ \ x_5 &&\leq 4\\
&\hspace{5.3cm} \vec x &&\geq \vec 0
\end{alignat*}

* (4 points) Write the LP dual of this problem.
* (3 points) Use CVXPY to compute the primal and dual optimum solutions and compare their values.
* (3 points) Check the complementary slackness conditions.

# Problem 1.

Consider the following LP problem:

\begin{alignat*}{2}
\max \ &4x_1-2x_2+5x_3+6x_4+7x_5\\
\\
\textrm{s.t} \quad
&2x_1 + 2x_2 - 4x_3 + 4x_4 + 8x_5 &&\leq 6\\
&2x_1 + \ \ {}x_2 - 2x_3 - \ \ x_4 - 3x_5 &&\geq -1\\
&5x_1 - 2x_2 + 4x_3 + 4x_4 + 2x_5 &&= 5\\
&2x_1 - 2x_2 + 5x_3 + 3x_4 + \ \ x_5 &&\leq 4\\
&\hspace{5.3cm} \vec x &&\geq \vec 0
\end{alignat*}

* (4 points) Write the LP dual of this problem.
* (3 points) Use CVXPY to compute the primal and dual optimum solutions and compare their values.
* (3 points) Check the complementary slackness conditions.

\begin{alignat*}{2}
\max \ &4x_1-2x_2+5x_3+6x_4+7x_5\\
\\
\textrm{s.t} \quad
&2x_1 + 2x_2 - 4x_3 + 4x_4 + 8x_5 &&\leq 6\\
&2x_1 + \ \ {}x_2 - 2x_3 - \ \ x_4 - 3x_5 &&\geq -1\\
&5x_1 - 2x_2 + 4x_3 + 4x_4 + 2x_5 &&= 5\\
&2x_1 - 2x_2 + 5x_3 + 3x_4 + \ \ x_5 &&\leq 4\\
&\hspace{5.3cm} \vec x &&\geq \vec 0
\end{alignat*}

$$
C^t =
\begin{bmatrix}
4 & -2 & 5 & 6 & 7
\end{bmatrix}
$$

$$
C =
\begin{bmatrix}
4 \\
-2 \\
5 \\
6 \\
7
\end{bmatrix}
$$

$$
B =
\begin{bmatrix}
6 \\
-1 \\
5 \\
4
\end{bmatrix}
$$ 

$$
A =
\begin{bmatrix}
2 & 2 & -4 & 4 & 8 \\
2 & 1 & -2 & -1 & -3 \\
5 & -2 & 4 & 4 & 2 \\
2 & -2 & 5 & 3 & 1 
\end{bmatrix}
$$

$$
A^t =
\begin{bmatrix}
2 & 2 & 5 & 2  \\
2 & 1 & -2 & -2 \\
-4 & -2 & 4 & 5\\
4 & -1 & 4 & 3  \\
8 & -3 & 2 & 1
\end{bmatrix}
$$

$$
O =
\begin{bmatrix}
\leq \\
\geq \\
= \\
\leq
\end{bmatrix}
$$ 

$$
O' =
\begin{bmatrix}
\geq \\
\geq \\
\geq \\
\geq
\end{bmatrix}
$$ 

primal = Maximize $C^t x \\$
        subject to $Ax O b \\$
        $ \vec x \geq \vec 0$

Dual = Minimize $ b^t y \\$
        subject to $A^t O' c \\$
        $y_1 \geq 0 , y_2 \leq 0, y_4 \geq 0$

The dual = 
\begin{alignat*}{2}
\min \ &6y_1-1y_2+5y_3+4y_4\\
\\
\textrm{s.t} \quad
&2y_1 + 2y_2 + 5y_3 + 2y_4 &&\geq 4\\
&2y_1 + \ \ {}y_2 - 2y_3 -2y_4 &&\geq -2\\
&-4y_1 - 2y_2 + 4y_3 + 5y_4 &&\geq 5\\
&4y_1 - 1y_2 + 4y_3 + 3y_4 &&\geq 6\\
&8y_1 - 3y_2 + 2y_3 + y_4 &&\geq 7\\
&y_1 \geq 0 , y_2 \leq 0, y_4 \geq 0
\end{alignat*}

In [19]:
import cvxpy as cp 
import numpy as np


#______PRIMAL_______
# Variables
x = cp.Variable(5)

# Objective function
c = [4, -2, 5, 6, 7]
b = [6, -1, 5, 4]
obj = cp.Maximize(c @ x)

# Constraints
A = np.array([[2,2,-4,4,8],
              [2,1,-2,-1,-3],
              [5,-2,4,4,2],
              [2,-2,5,3,1]])

Ax = A @ x

constraints_p = []
constraints_p.append(Ax[0] <= b[0])
constraints_p.append(Ax[1] >= b[1])
constraints_p.append(Ax[2] == b[2])
constraints_p.append(Ax[3] <= b[3])
constraints_p.append(x >= 0)


# Solution
prob = cp.Problem(obj, constraints_p)
opt = prob.solve()
print(f"Objective value: {round(opt, 2)}")
print(f"Solution vector: {[round(float(i.value), 2) for i in x]}")


#______DUAL_______
# Variables
y = cp.Variable(4)

# Objective function
b_t = np.transpose(b)
obj_d = cp.Minimize(b_t @ y)

A_t = np.transpose(A)
Ay = A_t @ y

constraints_d = [Ay >= c]
constraints_d.append(y[0] >= 0)
constraints_d.append(y[1] <= 0)
constraints_d.append(y[3] >= 0)
# Solution
prob_d = cp.Problem(obj_d, constraints_d)
opt_d = prob_d.solve()
print(f"Objective value: {round(opt_d, 2)}")
print(f"Solution vector: {[round(float(i.value), 2) for i in y]}")

Objective value: 9.22
Solution vector: [0.63, 2.81, 1.54, 0.0, 0.66]
Objective value: 9.22
Solution vector: [0.53, -0.37, 0.34, 1.0]


The complementry slackness conditions

If $(x*,y*)$ are a pair of optimal solutions fo the primal and dual, then:
* $x_i* = 0$ or corresponding constraint in the dual satisfied with equality ("tight")
* $y_j* = 0$ or corresponding constraint in the primal satisfied with equality ("tight")

These conditions are necessary and sufficient for optimality

$x_1 = 0.63 \implies 2y_1 + 2y_2 + 5y_3 + 2y_4 = 4 \\$
$x_2 = 2.81 \implies 2y_1 + 1y_2 - 2y_3 - 2y_4 = -2\\$
$x_3 = 1.54 \implies -4y_1 - 2y_2 + 4y_3 + 5y_4 = 5\\$
$x_5 = 0.66 \implies 8y_1 - 3y_2 + 2y_3 + y_4 = 7\\$

$y_1 = 0.53 \implies 2x_1 + 2x_2 - 4x_3 + 4x_4 + 8x_5 = 6\\$
$y_2 = -0.37 \implies 2x_1 + x_2 - 2x_3 - x_4 - 3x_5 = -1\\$
$y_3 = 0.34 \implies 5x_1 - 2x_2 + 4x_3 + 4x_4 + 2x_5 = 5\\$
$y_4 = 1.0 \implies 2x_1 - 2x_2 + 5x_3 + 3x_4 + x_5 = 4\\$

Lets check if these conditions are true, we substitue the variable with their actual value

$2(0.53) + 2(-0.37) + 5(0.34) + 2(1.0) = VALUE \approx 4 \\$
$2(0.53) + 1(-0.37) - 2(0.34) - 2(1.0) = VALUE\approx -2 \\$
$-4(0.53) - 2(-0.37) + 4(0.34) + 5(1.0) = VALUE\approx 5 \\$
$8(0.53) - 3(-0.37) + 2(0.34) + 1(1.0) = VALUE\approx 7 \\$

$2(0.63) + 2(2.81) - 4(1.54) + 4x_4 + 8(0.66) = VALUE \approx 6 \\$
$2(0.63) + 1(2.81) - 2(1.54) - x_4 - 3(0.66) = VALUE\approx -1 \\$
$5(0.63) - 2(2.81) + 4(1.54) + 4x_4 + 2(0.66) = VALUE \approx 5 \\$
$2(0.63) - 2(2.81) + 5(1.54) + 3x_4 + 1(0.66) = VALUE\approx 4 \\$


# Problem 2.

Consdier the LP problem:
\begin{alignat*}{2}
\max \ &6x_1 - 5x_3\\
\\
\textrm{s.t} \quad
&6x_1 - 3x_2 + x_3 &&= 2\\
&3x_1 + 4x_2 + x_3 &&\leq 5\\
&\ \ x_1 - 7x_2 &&\leq 5\\
&\hspace{2.45cm} x_1 &&\geq 0\\
&\hspace{2.45cm} x_2 &&\leq 0\\
&\hspace{2.45cm} x_3 &&\text{ unrestricted}
\end{alignat*}

* (3 points) Write the LP dual of this problem.
* (4 points) Consider the feasible solution $\vec x = (0,0,2)$ to the primal. Check if this solution is optimal by using the complementary slackness conditions to write down the corresponding dual solution.
* (3 points) Use complementary slackness to check if the primal feasible solution $\vec x = (1,0,-4)$ is optimal.


$$
C^t =
\begin{bmatrix}
6 & 0 & -5
\end{bmatrix}
$$

$$
C =
\begin{bmatrix}
6 \\
0 \\
-5
\end{bmatrix}
$$

$$
B =
\begin{bmatrix}
2 \\
-5 \\
5
\end{bmatrix}
$$ 

$$
A =
\begin{bmatrix}
6 & -3 & 1\\
3 & 4 & 1 \\
1 & -7 & 0
\end{bmatrix}
$$

$$
A^t =
\begin{bmatrix}
6 & 3 & 1\\
-3 & 4 & -7\\
1 & 1 & 0
\end{bmatrix}
$$

$$
O =
\begin{bmatrix}
= \\
\leq\\
\leq
\end{bmatrix}
$$ 

$$
O' =
\begin{bmatrix}
\geq \\
\leq \\
=
\end{bmatrix}
$$ 

primal = Maximize $C^t x \\$
        subject to $Ax O b \\$
        $x_1 \geq 0, x_2 \leq 0, x_3 unrestricted$

Dual = Minimize $ b^t y \\$
        subject to $A^t O' c \\$
        $y_2 \geq 0 , y_3 \geq 0$

The dual = 
\begin{alignat*}{2}
\min \ &2y_1-5y_2+5y_3\\
\\
\textrm{s.t} \quad
&6y_1 + 3y_2 + y_3 && \geq 6\\
&-3y_1 + 4y_2 - 7y_3 &&\leq 0\\
&1y_1 + 1y_2 + 0y_3&& = -5 \\
&y_2 \geq 0 , y_3 \geq 0
\end{alignat*}

b)

let's check if the primal feasible solution $\vec x = (0,0,2)$ is optimal. The complementary slackness condition tells us that: $$6y_1 + 3y_2 + y_3 \geq 6 $$ $$-3y_1 + 4y_2 - 7y_3 \leq 0$$ $$1y_1 + 1y_2 + 0y_3 = -5$$


c)

Now let's check if the primal feasible solution $\vec x = (1,0,-4)$ is optimal. The complementary slackness condition tells us that: $$6y_1 + 3y_2 + y_3 = 6 $$ $$-3y_1 + 4y_2 - 7y_3 \leq 0$$ $$1y_1 + 1y_2 + 0y_3 = -5$$

# Problem 3.

Consider the primal-dual algorithm for vertex cover discussed in class.
* (4 points) Run the algorithm by hand on the graph in the figure below (from your previous homework). Show the values of the primal and dual variables at each iteration.
* (6 points) Implement the primal-dual algorithm as a python script to compute (approximate) vertex covers and run it for the random graph $G(100,0.1)$ from the previous homework.

<img src="https://tinyurl.com/tsnuz2c" alt="Drawing" style="width: 180px;"/>

If the image does not load, try the direct link: https://tinyurl.com/tsnuz2c